src/OpenTelemetry.Instrumentation.ConfluentKafka/InstrumentedConsumer.cs

```diff
@@ -0,0 +1,330 @@
+    {
+        return consumer.QueryWatermarkOffsets(topicPartition, timeout);
+    }
+
+    public void Close()
+    {
+        consumer.Close();
+    }
+
+    private static PropagationContext ExtractPropagationContext(Headers headers)
+        => Propagators.DefaultTextMapPropagator.Extract(default, headers, ExtractTraceContext);
+
+    private static IEnumerable<string> ExtractTraceContext(Headers headers, string value)
+    {
+        if (headers?.TryGetLastBytes(value, out var bytes) == true)
+        {
+            yield return Encoding.UTF8.GetString(bytes);
+        }
+    }
+
+    private void InstrumentConsumption(DateTimeOffset startTime, ConsumeResult<TKey, TValue> result)
+    {
+        ConsumeResult consumeResult = new(result.TopicPartitionOffset, result.Message.Headers, result.Message.Key);
+        PropagationContext propagationContext = consumeResult.Headers != null
+            ? ExtractPropagationContext(consumeResult.Headers)
+            : default;
+        var spanName = string.IsNullOrEmpty(consumeResult.TopicPartitionOffset?.Topic)
+            ? OperationName
+            : string.Concat(consumeResult.TopicPartitionOffset!.Topic, " ", OperationName);
+
+        using var activity = this.StartReceiveActivity(spanName, propagationContext, startTime, consumeResult.TopicPartitionOffset, consumeResult.Key);
+        activity?.Stop();
+    }
+
+    private void InstrumentConsumption(DateTimeOffset startTime, ConsumeException exception)
+    {
+        ConsumeResult consumeResult = new(exception.ConsumerRecord.TopicPartitionOffset, exception.ConsumerRecord.Message.Headers, exception.ConsumerRecord.Message.Key);
+        PropagationContext propagationContext = consumeResult.Headers != null
+            ? ExtractPropagationContext(consumeResult.Headers)
+            : default;
+        var spanName = string.IsNullOrEmpty(consumeResult.TopicPartitionOffset?.Topic)
+            ? OperationName
+            : string.Concat(consumeResult.TopicPartitionOffset!.Topic, " ", OperationName);
+
+        using var activity = this.StartReceiveActivity(spanName, propagationContext, startTime, consumeResult.TopicPartitionOffset, consumeResult.Key);
+        activity?.Stop();
+    }
+
+    private Activity? StartReceiveActivity(string spanName, PropagationContext propagationContext, DateTimeOffset start, TopicPartitionOffset? topicPartitionOffset, object? key)
+    {
+        ActivityLink[] activityLinks = propagationContext.ActivityContext.IsValid()
+            ? new[] { new ActivityLink(propagationContext.ActivityContext) }
+            : Array.Empty<ActivityLink>();
+
+        ActivityContext activityContext = Activity.Current is null
+            ? propagationContext.ActivityContext
+            : default;
+
+        Activity? activity = ConfluentKafkaCommon.ActivitySource.StartActivity(spanName, kind: ActivityKind.Consumer, links: activityLinks, startTime: start, parentContext: activityContext);
```

Comment:
The sample table here https://github.com/open-telemetry/semantic-conventions/blob/main/docs/messaging/kafka.md#apache-kafka-with-quarkus-or-spring-boot-example states the opposite :).
Please advise @lachmatt / @pyohannes / @lmolkova / @Kielek

Reply:
In case of receive activity, parenting is problematic.
If we could, we should start this activity before receiving starts - we can't do it today because links have to be added at the beginning. But otel spec now allows to add links after span creation and it's a matter of time for .NET to support it.
Starting activity at the beginning allows to correlate any nested spans or logs with it. It won't be possible with the current approach.


For the processing callback instrumentation, having both - links and parenting is just fine. 

I'll file an issue on the example on the semconv - it's not up to date.

Reply:
@lmolkova The .NET Kafka SDK exposes a `Subscribe` + `Consume` API that forces a caller to first call `Subscribe` and then call `Consume` in a loop. My understanding of the messaging semantic convention is that each call to `Consume` (which can return 0 or 1 message) that returns 1 message should create a `receive` span. Can you confirm ?

Reply:
@lachmatt I kept the link but removed the parent/child relation